### PR TITLE
Code and patterns improvements

### LINKED ISSUE #35: Wrong correction of  ctg(x) and arctg(x)
Function `replace_incorrect_functions` in `math.math_function.py` works wrong. It replace `ctg` to `ctan` and `arcctg` to `arcctan`.
image

### PR DESCRIPTION
**Pull request by Korostast**
**Code cleaning up, patterns update and bugfixes**

- Main features:
   - Update patterns. They are more flexible and reliable (I hope)
   - If ParseError was raised while processing the request, then we try to correct words and repeat processing
   - Plotting is limited by number of functions (see constant FUNCTIONS_COUNT in GraphParser)
- Bugfixes:
   - Fix: infinities sometimes were determined as vertical asymptotes
   - Fix: Sympy don't want to find real solutions for equations like "2^x - 4 = 0"
   - Fix: incorrect replacements of some trigonometric functions (e.g. arcctg -> arcctan)
   - Partial fix: bot doesn't find vertical asymptotes of periodic functions (or partly periodic functions)
- Code improvements:
   - Fix #35
   - Several comments corrections
   - Some code refactoring and simplification
   - Add type hints for some function that doesn't have hints or have incorrect hints
- Others:
   - .pylintrc corrections

### CODE REVIEW

File: source/math/math_function.py [hunk 1]
```diff
@@ -188,6 +189,13 @@ def is_odd(self, *symbols) -> bool:
 
         return even_func == sy.simplify(function * (-1))
 
+    def _check_v_asymptote(self, symbol, point) -> bool:
+        left_limit = sy.limit(self.simplified_expr, symbol, point, '+')
+        right_limit = sy.limit(self.simplified_expr, symbol, point, '-')
+        if not point.is_infinite and (left_limit.is_infinite or right_limit.is_infinite):
+            return True
+        return False
```
Comment (RaEzhov) [inline source/math/math_function.py:197]:
```suggestion
        return not point.is_infinite and (left_limit.is_infinite or right_limit.is_infinite)
```

File: source/math/calculus_parser.py [hunk 2]
```diff
@@ -296,14 +297,19 @@ def parse(self, query: str) -> bool:
...
-            pattern_dict = json.load(file)
+        pattern_dict = Config.analysis_patterns
 
         # Check if input match any pattern
-        is_pattern_found = self._find_pattern(query, pattern_dict, False)
-        if is_pattern_found:
-            return True
+        try:
+            is_pattern_found = self._find_pattern(query, pattern_dict, False)
+            if is_pattern_found:
```
Comment (RaEzhov) [inline source/math/calculus_parser.py:305]:
```suggestion
            if self._find_pattern(query, pattern_dict, False):
```

File: source/math/calculus_parser.py [hunk 3]
```diff
@@ -296,14 +297,19 @@ def parse(self, query: str) -> bool:
...
+                return True
+        except ParseError as err:
+            # Maybe we should correct some words here. If nothing was changed, then we throw previous exception
+            is_pattern_found = self._find_pattern(query, pattern_dict, True)
+            if is_pattern_found:
```
Comment (RaEzhov) [inline source/math/calculus_parser.py:310]:
```suggestion
            if self._find_pattern(query, pattern_dict, True):
```

**Review by RaEzhov — APPROVED** [review]
Good job

File: source/math/math_function.py (hunk 1 above)
Comment (Korostast) [reply source/math/math_function.py:197]: Yes,  good suggestion

**Pull request merged**